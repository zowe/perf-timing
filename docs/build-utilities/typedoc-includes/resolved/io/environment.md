<!-- Fix needs sent to typedoc to allow for sub includes. For now a direct link will do-->