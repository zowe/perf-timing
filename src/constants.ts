/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

/**
 * This is the environmental prefix that will be added to any configuration
 * settings for performance.
 *
 * @external
 */
export const ENV_PREFIX = "PERF_TIMING";


/**
 * Symbol used to ensure uniqueness across all possible instances of this package.
 *
 * @internal
 */
export const GLOBAL_SYMBOL = Symbol.for("org.zowe.perf-timing");
