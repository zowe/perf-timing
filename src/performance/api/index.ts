export * from "./errors";
export * from "./interfaces";
export * from "./PerformanceApi";
