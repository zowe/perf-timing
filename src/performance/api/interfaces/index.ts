export * from "./IFunctionMetric";
export * from "./IFunctionTimer";
export * from "./IMeasurement";
export * from "./IMeasurementMetric";
export * from "./IMeasureTimer";
export * from "./IMetrics";
