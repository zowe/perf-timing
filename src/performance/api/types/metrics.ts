/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IMeasurementEntry, IMetric, IPerformanceEntry } from "../interfaces";

/**
 * Data collected for all functions that were monitored with the
 * {@link PerformanceApi.watch} function.
 */
export type FunctionMetric = IMetric<IPerformanceEntry>;

/**
 * Data collected for all measurements taken through the
 * {@link PerformanceApi.measure} function.
 */
export type MeasurementMetric = IMetric<IMeasurementEntry>;
