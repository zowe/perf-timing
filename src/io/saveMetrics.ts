/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { ENV_PREFIX } from "../constants";
import { Environment } from "../environment";
import * as fs from "fs-extra";
import * as os from "os";
import { IPerformanceMetrics } from "../performance/manager";

/**
 * Environment key prefix for the max history value. Max history refers to the
 * max number of log entries to keep when performance is enabled.
 *
 * @external
 */
const ENV_IO_MAX_HISTORY = `${ENV_PREFIX}_IO_MAX_HISTORY`;

/**
 * Environment key prefix for where the logs are saved.
 *
 * @external
 */
const ENV_IO_SAVE_DIR = `${ENV_PREFIX}_IO_SAVE_DIR`;

// tslint:disable:no-magic-numbers
Environment
    .register(ENV_IO_MAX_HISTORY, 5)
    .register(ENV_IO_SAVE_DIR, `${os.homedir()}/.perf-timing`);
// tslint:enable:no-magic-numbers

/**
 * Save metrics into the next log file.
 *
 * Upon execution this function will first roll any logs based on the value
 * present within environmental {@link ENV_IO_MAX_HISTORY} variable. The directory
 * that is checked will be determined from the {@link ENV_IO_SAVE_DIR} value.
 *
 * @param data The object data to save in a file.
 *
 * @todo This function needs to be enhanced so that an environment variable can
 * influence the file name.
 *
 * @internal
 */
export function saveMetrics(data: IPerformanceMetrics) {
    // @FUTURE This is where file save location will be determined by
    // @FUTURE environment settings for aggregation at a later time.
    const directory = Environment.getValue(ENV_IO_SAVE_DIR);

    fs.ensureDirSync(directory);

    // Prior to saving roll any metrics beforehand
    // Get the starting index
    let historyIndex = Environment.getNumber(ENV_IO_MAX_HISTORY);

    const lastFile = getMetricFileName(directory, historyIndex);

    // Delete the last file when history limit has been reached
    fs.removeSync(lastFile);

    // Move all historical logs down by 1
    while (--historyIndex > 0) {
        const fileName = getMetricFileName(directory, historyIndex);

        if (fs.existsSync(fileName)) {
            fs.renameSync(fileName, getMetricFileName(directory, historyIndex + 1));
        }
    }

    // Write the first metric
    fs.writeFileSync(getMetricFileName(directory, 1), JSON.stringify(data));
}

/**
 * Get a constant file path for a given history item index.
 *
 * @param directory The directory of the file
 * @param index The index of the file in the history
 *
 * @returns The formatted file name that will be saved
 *
 * @internal
 */
function getMetricFileName(directory: string, index: number) {
    return `${directory}/metrics.${index}.json`;
}
