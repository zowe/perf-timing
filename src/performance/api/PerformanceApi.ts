/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IPerformanceApi, IPerformanceApiManager } from "../manager/interfaces";
import {
    ICollectionObserver,
    IFunctionObserver,
    IMeasurementObserver,
    IMetric,
    IMetrics,
    INodeTiming,
    IPerformanceEntry,
    ISystemInformation
} from "./interfaces";

/**
 * A type that generically defines all collection observer maps that are defined
 * to the Performance API.
 *
 * @param T The type of observer stored in the map. The observer must be an
 *          {@link ICollectionObserver} type with entries adhering to the
 *          {@link IPerformanceEntry} interface.
 *
 * @internal
 */
type CollectionMap<T extends ICollectionObserver<IPerformanceEntry>> = Map<string, T>;

/**
 * The underlying api that provides hooks into
 * {@link https://nodejs.org/dist/latest-v10.x/docs/api/perf_hooks.html Node's Performance Timing APIs}.
 *
 * Use this class in place of relying on Node's experimental APIs.
 */
export class PerformanceApi implements IPerformanceApi {

    /**
     * Internal method for getting the errors that can be thrown. It is smart
     * and will only do the require once and cache the results for later calls.
     *
     * @internal
     */
    private static get _errors(): typeof PerformanceApi._errorImport {
        if (!PerformanceApi._errorImport) {
            PerformanceApi._errorImport = require("./errors");
        }

        return PerformanceApi._errorImport;
    }

    /**
     * Cache of loaded error objects. Only populated on first call to {@link PerformanceApi._errors}
     *
     * @internal
     */
    private static _errorImport: typeof import("./errors");

    /**
     * Aggregate all entries present in a {@link CollectionMap} into an array output.
     *
     * Each item present in the map will correspond to a single entry in the final array output. The
     * strings will now become the name property and each data point will be stored in the final
     * object. The statistical analysis present is generated from the IPerformanceEntry instances
     * present in the entries portion of that element.
     *
     *
     * @param map A {@link CollectionMap} of observers containing entries to process.
     * @param T The type representing the raw data stored in the observer. It is important that this
     *          data extends {@link IPerformanceEntry} for proper data processing.
     *
     * @returns An array of {@link IMetric} data points representing each key/value present in the
     *          CollectionMap.
     *
     * @internal
     */
    private static _aggregateData<T extends IPerformanceEntry>(map: CollectionMap<ICollectionObserver<T>>): Array<IMetric<T>> {
        // Get all observers present within the map.
        const observers = map.entries();
        const output: Array<IMetric<T>> = [];

        // Loop through each observer in the map
        for (const [name, observer] of observers) {
            let totalDuration = 0;

            // Calculate the total duration represented by all entries
            for (const entry of observer.entries) {
                totalDuration += entry.duration;
            }

            // Create the data point for the map using the key as the name
            output.push({
                name,
                calls: observer.entries.length,
                totalDuration,
                averageDuration: totalDuration / observer.entries.length,
                entries: observer.entries
            });
        }

        return output;
    }

    /**
     * Created for consistent reference to Node Performance Timing API.
     *
     * @see {@link PerformanceApi.watch}
     */
    public timerify = this.watch;

    /**
     * Created for consistent reference to Node Performance Timing API.
     *
     * @see {@link PerformanceApi.unwatch}
     */
    public untimerify = this.unwatch;

    /**
     * Internal map of all created function timers. The key represents the name
     * and the value is an {@link IFunctionObserver} instance.
     *
     * @internal
     */
    private _functionObservers: CollectionMap<IFunctionObserver> = new Map();

    /**
     * Internal map of all created measurement timers. The key represents the name
     * and the value is an {@link IMeasurementObserver} instance.
     *
     * @internal
     */
    private _measurementObservers: CollectionMap<IMeasurementObserver> = new Map();

    /**
     * This variable holds the import from the Node JS performance hooks
     * library.
     *
     * @internal
     */
    private readonly _perfHooks: typeof import("perf_hooks");

    /**
     * Construct the performance API. The constructor is not intended to be used
     * by any class except for the {@link PerformanceApiManager} of this class.
     *
     * @internal
     *
     * @param _manager The manager of this API class.
     */
    constructor(private readonly _manager: IPerformanceApiManager) {
        // Check if performance utilities should be enabled.
        if (this._manager.isEnabled) {
            // Delay the require so we don't waste resources when performance
            // isn't needed.
            this._perfHooks = require("perf_hooks");
        }
    }

    /**
     * Clears marks created using {@link mark}.
     *
     * This method does nothing if the Performance API is not enabled. It also takes advantage
     * of the namespace applied within mark; however, if name is not specified, all marks
     * will be cleared from all package.
     *
     * @param name The name of the mark to clear. If not specified, all marks are
     *             cleared.
     *
     * @see {@link https://nodejs.org/api/perf_hooks.html#perf_hooks_performance_clearmarks_name clearMarks()}
     */
    public clearMarks(name?: string) {
        if (this._manager.isEnabled) {
            // Add the namespace to the mark for uniqueness across packages
            this._perfHooks.performance.clearMarks(
                name ? this._addPackageNamespace(name) : undefined
            );
        }
    }

     /**
      * Aggregate metrics that have been captured for this instance of the API
      * and return them back in a data format.
      *
      * @returns An object representing metrics that have been prepared to be
      *          written to a file.
      *
      * @throws {@link PerformanceNotCapturedError} when the method is called and performance captures
      *                                             are not enabled.
      */
    public getMetrics(): IMetrics {
        if (this._manager.isEnabled) {
            // @TODO All metrics should be stopped before reporting
            return {
                functions: PerformanceApi._aggregateData(this._functionObservers),
                measurements: PerformanceApi._aggregateData(this._measurementObservers)
            };
        }
        else {
            throw new PerformanceApi._errors.PerformanceNotCapturedError();
        }
    }

    /**
     * Get the node timing information provided by the underlying node APIs.
     *
     * @returns The node timing information.
     *
     * @throws {@link PerformanceNotCapturedError} when the method is called and performance captures
     *                                             are not enabled.
     */
    public getNodeTiming(): INodeTiming {
        if (this._manager.isEnabled) {
            const timing = this._perfHooks.performance.nodeTiming;

            return {
                bootstrapComplete: timing.bootstrapComplete,
                duration: timing.duration,
                loopStart: timing.loopStart,
                loopExit: timing.loopExit,
                name: timing.name,
                nodeStart: timing.nodeStart,
                startTime: timing.startTime,
                v8Start: timing.v8Start
            };
        } else {
            throw new PerformanceApi._errors.PerformanceNotCapturedError();
        }
    }

    /**
     * Gathers and returns valuable system information to help understand the environment
     * where the data was captured.
     *
     * @returns System information formatted in an object ready to write to a file.
     *
     * @throws {@link PerformanceNotCapturedError} when the method is called and performance captures
     *                                             are not enabled.
     */
    public getSysInfo(): ISystemInformation {
        if (this._manager.isEnabled) {
            // Dynamically import os so that this process doesn't have to bother
            // if performance is not enabled.
            const os: typeof import("os") = require("os");

            const freeMem = os.freemem();
            const totalMem = os.totalmem();

            return {
                argv: process.argv,
                cpus: os.cpus(),
                loadavg: os.loadavg(), // On windows this will always be [0, 0, 0]
                memory: {
                    free: freeMem,
                    total: totalMem,
                    usage: totalMem - freeMem,
                    usagePercentage: ((totalMem - freeMem) / totalMem) * 100 // tslint:disable-line:no-magic-numbers
                },
                network: {
                    hostname: os.hostname(),
                    interfaces: os.networkInterfaces()
                },
                os: `${os.type()} ${os.arch()} ${os.release()}`,
                platform: os.platform(),
                shell: os.userInfo().shell,
                uptime: os.uptime()
            };
        } else {
            throw new PerformanceApi._errors.PerformanceNotCapturedError();
        }
    }

    public mark(name: string) {
        if (this._manager.isEnabled) {
            this._perfHooks.performance.mark(
                this._addPackageNamespace(name)
            );
        }
    }


    // @TODO document
    public measure(name: string, startMark: string, endMark: string) {
        if (this._manager.isEnabled) {
            let mapObject: IMeasurementObserver;

            if (this._measurementObservers.has(name)) {
                mapObject = this._measurementObservers.get(name);
            } else {
                // Create the map object when needed
                mapObject = {
                    observer: undefined,
                    isConnected: false,
                    entries: []
                };

                this._measurementObservers.set(name, mapObject);
            }

            // Remap the name to the unique package name. This will not be visible
            // in the output of the command, but it will allow 2 instances of this
            // package to not conflict accidentally in their naming.
            const namespaceName = this._addPackageNamespace(name);
            const namespaceStartMark = this._addPackageNamespace(startMark);
            const namespaceEndMark = this._addPackageNamespace(endMark);

            // If the map object is already connected, we don't need to do anything.
            // The metric will be picked up by the for loop in the observer.
            if (!mapObject.isConnected) {
                // Create the observer that will capture the metrics.
                mapObject.observer = new this._perfHooks.PerformanceObserver((list) => {
                    const entries = list.getEntriesByName(namespaceName);

                    // Loop through each entry and save the metrics
                    if (entries.length > 0) {
                        for (const entry of entries) {
                            mapObject.entries.push({
                                name: entry.name,
                                startTime: entry.startTime,
                                duration: entry.duration,
                                startMarkName: namespaceStartMark,
                                endMarkName: namespaceEndMark
                            });
                        }

                        mapObject.observer.disconnect();
                        mapObject.isConnected = false;
                    }
                });
                mapObject.observer.observe({ entryTypes: ["measure"], buffered: true });
                mapObject.isConnected = true;
            }

            this._perfHooks.performance.measure(namespaceName, namespaceStartMark, namespaceEndMark);
        }
    }

    public unwatch(fn: ((...args: any[]) => any), name?: string) {
        if (this._manager.isEnabled) {
            let timer = fn.name;

            // Extract the name of the function if necessary
            if (name != null) {
                timer = name;
            } else {
                // When timerifing(?) functions, node will prepend the function
                // name with timerified . This regex will strip out that name
                timer = /timerified (.*)/g.exec(timer)[1];
            }

            const timerRef = this._functionObservers.get(timer);

            if (timerRef !== undefined) {
                timerRef.observer.disconnect();
                timerRef.isConnected = false;
                return timerRef.originalFunction;
            } else {
                throw new PerformanceApi._errors.TimerDoesNotExistError(timer);
            }
        } else {
            return fn;
        }
    }

    public watch(fn: (...args: any[]) => any, name?: string) {
        if (this._manager.isEnabled) {
            // Check if we should use the function name or the passed name for tracking.
            if (name == null) {
                name = fn.name;
            }

            // @TODO allow for a name to be reconnected if the observer has been disconnected.
            // Throw an error if the timer already exists in the map.
            if (this._functionObservers.has(name)) {
                throw new PerformanceApi._errors.TimerNameConflictError(name);
            }

            // Create the observer and store it in the map.
            const observerObject: IFunctionObserver = {
                observer: undefined,
                isConnected: false,
                entries: [],
                originalFunction: fn
            };

            this._functionObservers.set(name, observerObject);

            // Create a function observer
            observerObject.observer = new this._perfHooks.PerformanceObserver((list) => {
                const entries = list.getEntriesByName(fn.name);

                for (const entry of entries) {
                    observerObject.entries.push(entry);
                }
            });

            observerObject.observer.observe({ entryTypes: ["function"], buffered: true });
            observerObject.isConnected = true;

            // Wrap the function in a timer
            return this._perfHooks.performance.timerify(fn);
        } else {
            // If performance is not enabled, we need to return the function
            // so original code doesn't get broken.
            return fn;
        }
    }

    private _addPackageNamespace(name: string): string {
        return `${this._manager.packageUUID}: ${name}`;
    }
}
