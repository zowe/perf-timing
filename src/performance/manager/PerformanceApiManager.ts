import { IPerformanceApi, IPerformanceApiManager } from "./interfaces";
import { PerformanceApi } from "../api/PerformanceApi";
import * as pkgUp from "pkg-up";
import { isArray } from "util";
import { ENV_PREFIX, GLOBAL_SYMBOL } from "../../constants";
import { Environment } from "../../environment";

//////////////////////////////////////////
/////////////// GLOBAL TYPING ////////////
//////////////////////////////////////////
// tslint:disable:no-namespace interface-name
// We need to add stuff to the global scope so that one package can manage
// all possible loaded packages. To do this, we need to disable a few rules.

/**
 * Because of how package dependencies work, there could be multiple instances
 * of this package within an application. To avoid each of these packages having
 * a process.exit hook, we add all of them to the global scope under a unique
 * symbol. The first instance created will be the manager of all of these global
 * packages. This allows for a single `process.on('exit')` hook to be registered.
 *
 * @see {@link PerformanceApiManager#_savePerformanceResults}
 */
declare namespace NodeJS {
    interface Global {
        [GLOBAL_SYMBOL]: Map<symbol, IPerformanceApi>;
    }
}

declare var global: NodeJS.Global;
// tslint:enable

/////////////////////////////// ///////////
/////////////// END TYPING ///////////////
//////////////////////////////////////////

/**
 * This class is a manager of all available performance tools
 */
export class PerformanceApiManager implements IPerformanceApiManager {
    // @TODO recommend wrapping stuff since we forward up requests through dummy object
    // @TODO to reduce overhead on getApi call
    public static readonly ENV_ENABLED = `${ENV_PREFIX}_ENABLED`;
    public static readonly ENV_ENABLED_KEY = "TRUE";
    public readonly isEnabled: boolean;
    public readonly packageUUID: string;

    private _instanceSymbol: symbol; // used to uniquely identify the instance

    private _managedApi: PerformanceApi;

    // @TODO Document
    constructor() {
        // First check if the environment prefix is set to the ENV_ENABLED_KEY value
        if (Environment.getValue(PerformanceApiManager.ENV_ENABLED).toUpperCase() === PerformanceApiManager.ENV_ENABLED_KEY) {
            // The environment was set so performance metrics are enabled.
            this.isEnabled = true;

            // Generate the packageUUID
            const pkg = require(pkgUp.sync());
            this.packageUUID = `${pkg.name}@${pkg.version}`;

            // Check if the global scope has been created. If it hasn't been created
            // this is the first manager called so it will become the main manager.
            if (!global[GLOBAL_SYMBOL]) {
                // Create the global map on first run
                global[GLOBAL_SYMBOL] = new Map();
                process.on("exit", () => this._savePerformanceResults());
            }
        } else {
            // Performance was not enabled.
            this.isEnabled = false;
        }
    }

    public getApi(): PerformanceApi {
        if (this._managedApi == null) {
            // Defers the import until it is needed, will improve performance when
            // performance api hasn't yet been called.
            const performanceApi: typeof PerformanceApi = require("../api/PerformanceApi").PerformanceApi;
            this._managedApi = new performanceApi(this);


            // Create a unique entry in the global map
            if (this.isEnabled) {
                this._instanceSymbol = Symbol(this.packageUUID);

                // Save the managedApi in the global space. It has been typed so that
                // changes to the wrapper methods of IPerformanceTools cannot be easily
                // changed.
                global[GLOBAL_SYMBOL].set(this._instanceSymbol, this._managedApi);
            }
        }

        // Can guarantee that this will be unique per package instance :)
        return this._managedApi;
    }

    private async _savePerformanceResults(): Promise<void> {
        const metrics = global[GLOBAL_SYMBOL].entries();
        const outputMetrics: any = {}; // @TODO proper typing

        // Get timing first to not skew the results
        outputMetrics.nodeTiming = this.getApi().getNodeTiming();

        for (const [key, value] of metrics) {
            const symbolValue = key.toString();

            // Place into an array to handle the case where the same
            // package might have existed twice.

            if (!isArray(outputMetrics[symbolValue])) {
                outputMetrics[symbolValue] = [];
            }

            outputMetrics[symbolValue].push(value.getMetrics());
        }

        console.dir(outputMetrics, {
            depth: 4,
            colors: true
        });

        // Require the IO utility at this point to reduce total number of
        // requires in the calling library when performance monitoring
        // is not enabled.
        (await import("../../io/saveMetrics")).saveMetrics(outputMetrics);

        // @TODO ensure there are no running timers before we print metrics
    }
}

// Register the PERF_TIMING_ENABLED variable
Environment.register(PerformanceApiManager.ENV_ENABLED, "");
