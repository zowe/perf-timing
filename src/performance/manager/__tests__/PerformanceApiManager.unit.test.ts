/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

jest.mock("../../api/PerformanceApi");
jest.mock("pkg-up");
jest.mock("../../../io");

import { PerformanceApi } from "../../api/PerformanceApi";
import { _GlobalType, PerformanceApiManager } from "../PerformanceApiManager";
import { GLOBAL_SYMBOL } from "../../../constants";
import * as pkgUp from "pkg-up";
import * as path from "path";
import { getMockWrapper } from "../../../../__tests__/utilities/types/MockWrapper";
import { saveMetrics } from "../../../io";

declare var global: _GlobalType;

describe("PerformanceApiManager", () => {
    // The total number of test json files available
    const NUM_JSON = 4;

    // Need to remember this so that we can replace the function after this test is done
    const processOnFunction = process.on;

    // Shorthand for getting the global symbol for purposes of this test.
    const getGlobal = () => global[GLOBAL_SYMBOL];

    // Shorthand for cleaning up after each test.
    const cleanup = () => {
        if (process.env[PerformanceApiManager.ENV_ENABLED]) {
            delete process.env[PerformanceApiManager.ENV_ENABLED];
        }

        if (global[GLOBAL_SYMBOL]) {
            delete global[GLOBAL_SYMBOL];
        }
    };

    /**
     * Calls the function registered to process.on("exit") by an api manager.
     * @param process The process mock instance where process.on exit was called
     * 
     * @returns the process call that is executed in the case that there might be a promise.
     */
    const triggerExit = (process: jest.Mock<any>) => {
        return process.mock.calls[0][1]();
    };

    /**
     * Get the path of a test json file.
     *
     * @param index The package.${index}.json value to get.
     * @returns a formulated path.
     */
    const getTestPath = (index: number) => path.join(__dirname, "test-json", `package.${index}.json`);

    /**
     * Sets the value of the environmental variable {@link PerformanceApiManager.ENV_ENABLED}
     * to the value passed.
     *
     * @param value The value to set to the Environmental variable.
     * @returns The value that was set
     */
    const setEnv = (value: string) => process.env[PerformanceApiManager.ENV_ENABLED] = value;

    beforeAll(() => {
        // Create a wrapper function to see what actually happens during a test call.
        (process as any).on = jest.fn();
    });

    afterAll(() => {
        (process as any).on = processOnFunction;
    });

    beforeEach(() => {
        jest.resetAllMocks();

        expect(getGlobal()).toBeUndefined();
    });

    afterEach(() => {
        cleanup();
    });

    it("should construct with performance disabled", () => {
        const testManager = new PerformanceApiManager();

        // Allows access to underlying private variables.
        const _testManager: any = testManager;

        expect(testManager.isEnabled).toBe(false);
        expect(testManager.packageUUID).toBeUndefined();
        expect(_testManager._api).toBeUndefined();

        // Test the API function
        expect(testManager.api).toBeInstanceOf(PerformanceApi);
        expect(PerformanceApi).toHaveBeenCalledWith(testManager);

        for (const testCase of ["Invalid", "FALSE", "FaLsE", ""]) {
            setEnv(testCase);
            expect((new PerformanceApiManager()).isEnabled).toBe(false);
        }
    });

    it("should construct with performance enabled", () => {
        setEnv(PerformanceApiManager.ENV_ENABLED_KEY);

        const testPath = getTestPath(1);
        const json = require(testPath);

        const mocks = getMockWrapper({
            pkgUpSync: pkgUp.sync,
            processOn: process.on
        });

        mocks.pkgUpSync.mockReturnValue(testPath);

        const testManager = new PerformanceApiManager();
        const _testManager: any = testManager;

        // Check class variables were properly set
        expect(testManager.isEnabled).toBe(true);
        expect(testManager.packageUUID).toBe(`${json.name}@${json.version}`);

        // Check that the global was created
        expect(getGlobal()).toBeInstanceOf(Map);

        // Check that the process.on exit was registered
        expect(mocks.processOn).toHaveBeenCalledTimes(1);
        expect(mocks.processOn).toHaveBeenCalledWith("exit", expect.any(Function));

        // Test that the right function was added
        _testManager._savePerformanceResults = jest.fn();
        triggerExit(mocks.processOn);
        expect(_testManager._savePerformanceResults).toHaveBeenCalled();

        // Test that the api works as expected
        expect(getGlobal().get(_testManager._instanceSymbol)).toBeUndefined();

        expect(testManager.api).toBeInstanceOf(PerformanceApi);
        expect(PerformanceApi).toHaveBeenCalledWith(testManager);

        // Check that this api was added to the global symbol
        expect(_testManager._instanceSymbol.toString()).toEqual(Symbol(testManager.packageUUID).toString());
        expect(getGlobal().get(_testManager._instanceSymbol)).toBe(testManager.api);

        // Now check that the capitalization doesn't affect if the utility gets enabled
        for(const testCase of ["TRUE", "true", "tRuE", "TRUe"]) {
            setEnv(testCase);
            expect((new PerformanceApiManager()).isEnabled).toBe(true);
        }
    });

    it("should default to the perf-timing package symbol when no package was found", () => {
        setEnv("true");

        const mocks = getMockWrapper({
            pkgUpSync: pkgUp.sync
        });

        const testPath = getTestPath(2);
        const testJson = require(testPath);

        mocks.pkgUpSync.mockReturnValueOnce(undefined).mockReturnValueOnce(testPath);

        const testManager = new PerformanceApiManager();

        expect(testManager.isEnabled).toBe(true);
        expect(testManager.packageUUID).toBe(`${testJson.name}@${testJson.version}`);

        expect(mocks.pkgUpSync).toHaveBeenCalledTimes(2);
        expect(mocks.pkgUpSync).toHaveBeenNthCalledWith(1);
        expect(mocks.pkgUpSync).toHaveBeenNthCalledWith(2, path.resolve(__dirname, "../"));
    });

    it("should create one api object", () => {
        const mocks = getMockWrapper({
            pkgUpSync: pkgUp.sync
        });

        // Check that the api is initialized only once for a manager
        const disabledManager = new PerformanceApiManager();

        expect(disabledManager.api).toBeInstanceOf(PerformanceApi);
        expect(disabledManager.api).toBeInstanceOf(PerformanceApi);
        expect(disabledManager.api).toBeInstanceOf(PerformanceApi);

        expect(PerformanceApi).toHaveBeenCalledTimes(1);

        jest.clearAllMocks();

        // Perform the same checks when on the main manager
        setEnv("true");
        mocks.pkgUpSync.mockReturnValueOnce(getTestPath(1));

        // Hook a jest fn in place of the map so we can easily check that set
        // was only called once.
        (global[GLOBAL_SYMBOL] as any) = {
            set: jest.fn()
        };

        const enabledManager = new PerformanceApiManager();

        expect(enabledManager.api).toBeInstanceOf(PerformanceApi);
        expect(enabledManager.api).toBeInstanceOf(PerformanceApi);
        expect(enabledManager.api).toBeInstanceOf(PerformanceApi);

        expect(PerformanceApi).toHaveBeenCalledTimes(1);

        expect(getGlobal().set).toHaveBeenCalledTimes(1);
        expect(getGlobal().set).toHaveBeenCalledWith((enabledManager as any)._instanceSymbol, (enabledManager as any)._api);

    });

    describe("ensure that only one class acts as the main manager.", () => {
        beforeEach(() => {
            setEnv("true");
        });

        it("should register only one metric gathering event", () => {
            const managerArray: PerformanceApiManager[] = [];

            const mocks = getMockWrapper({
                pkgUpSync: pkgUp.sync,
                processOn: process.on
            });

            const numTestJson = NUM_JSON;

            for(let i = 1; i <= numTestJson; i++) {
                const testPath = getTestPath(i);
                const testJson = require(testPath);

                mocks.pkgUpSync.mockReturnValueOnce(testPath);

                const manager = new PerformanceApiManager();

                managerArray.push(manager);

                expect(manager.packageUUID).toBe(`${testJson.name}@${testJson.version}`);
                expect(mocks.processOn).toHaveBeenCalledTimes(1);
            }

            // Ensure that the first manager created is the one that has the save called
            (managerArray[0] as any)._savePerformanceResults = jest.fn();
            triggerExit(mocks.processOn);
            expect((managerArray[0] as any)._savePerformanceResults).toHaveBeenCalledTimes(1);
        });

        it("should register manager's api in the global space when requested", () => {
            const numJson = NUM_JSON;
            const mocks = getMockWrapper({
                pkgUpSync: pkgUp.sync
            });
            const managers: PerformanceApiManager[] = [];

            // we loop more times than we have packages so that a package name@version
            // will be repeated a bit. This allows us to see that we don't have
            // possible clash problems.
            const testLoops = numJson + 2;

            // Initialize every manager
            for(let i = 0; i < testLoops; i++) {
                mocks.pkgUpSync.mockReturnValueOnce(getTestPath(i % numJson + 1));
                managers.push(new PerformanceApiManager());
            }

            // Now loop through each one and see if the api works (calling once will create the manager)
            for(const manager of managers) {
                expect(manager.api).toBeInstanceOf(PerformanceApi);
            }

            // Now that the apis are created, check that they were properly added to the global symbol
            for(const manager of managers) {
                expect(getGlobal().get((manager as any)._instanceSymbol)).toBe((manager as any)._api);
            }
        });
    });

    describe("savePerformanceResults", () => {
        beforeEach(() => {
            setEnv("true");
        });

        const testMap: Map<string, number> = new Map([
            ["should save when there is just one manager", 1],
            ["should save when there are multiple managers. (no name conflicts)", NUM_JSON],
            ["should save when there are multiple managers. (some name conflicts)", NUM_JSON + 2],
        ]);

        for(const [name, numLoops] of testMap) {
            it(name, async () => {
                const mocks = getMockWrapper({
                    pkgUpSync: pkgUp.sync,
                    processOn: process.on,
                    saveMetrics
                });

                for (let i = 0; i < numLoops; i++) {
                    mocks.pkgUpSync.mockReturnValueOnce(getTestPath(i % NUM_JSON + 1));

                    const manager = new PerformanceApiManager();
                    const managerApi = getMockWrapper(manager.api);

                    managerApi.getMetrics.mockReturnValue(`manager ${i}: manager.api.getMetrics()`);
                    managerApi.getNodeTiming.mockReturnValue(`manager ${i}: manager.api.getNodeTiming()`);
                    managerApi.getSysInfo.mockReturnValue(`manager ${i}: manager.api.getSysInfo()`);
                }

                // Simulate the process exit call
                await triggerExit(mocks.processOn);

                expect(mocks.saveMetrics).toHaveBeenCalledTimes(1);
                
                // Validate that the manager was the one that called the function
                expect(mocks.saveMetrics.mock.calls[0][0].nodeTiming).toContain("manager 0:");
                expect(mocks.saveMetrics.mock.calls[0][0].systemInformation).toContain("manager 0:");
                expect(mocks.saveMetrics.mock.calls[0][0]).toMatchSnapshot();
            });
        }
    });
});
