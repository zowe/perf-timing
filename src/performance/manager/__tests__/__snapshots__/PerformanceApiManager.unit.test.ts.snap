// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerformanceApiManager savePerformanceResults should save when there is just one manager 1`] = `
Object {
  "metrics": Object {
    "Symbol(package1@1.5.0)": Array [
      Object {
        "testGetMetricsReturn": "get metrics works",
      },
    ],
  },
  "nodeTiming": Object {
    "testNodeTiming": "node timing works",
  },
  "systemInformation": Object {
    "testSysInfoReturn": "sys info works",
  },
}
`;
